File: DFAChecker.java
package DFAChecker;

import java.util.*;
import java.io.*;

public class DFAChecker {
    private Set<Character> alphabet;
    private Set<Character> states;
    private char startState;
    private Set<Character> acceptStates;
    private Map<String, Character> transitions;

    public DFAChecker() {
        alphabet = new HashSet<>();
        states = new HashSet<>();
        acceptStates = new HashSet<>();
        transitions = new HashMap<>();
    }

    public void loadDFA(String filename) throws IOException {
        BufferedReader reader = new BufferedReader(new FileReader(filename));
        String line;
        int lineNum = 0;
        while ((line = reader.readLine()) != null) {
            lineNum++;
            switch (lineNum) {
                case 1: parseSet(line, alphabet); break;
                case 2: parseSet(line, states); break;
                case 3: startState = parseState(line); break;
                case 4: parseSet(line, acceptStates); break;
                default: parseTransition(line);
            }
        }
        reader.close();
    }

    private void parseSet(String line, Set<Character> set) {
        line = line.replaceAll("[{}\\s]", "");
        for (char c : line.toCharArray()) {
            set.add(c);
        }
    }

    private char parseState(String line) {
        return line.charAt(0);
    }

    private void parseTransition(String line) {
        String[] parts = line.split("->");
        String[] transition = parts[0].substring(1, parts[0].length() - 1).split(",");
        char fromState = transition[0].charAt(0);
        char symbol = transition[1].charAt(0);
        char toState = parts[1].charAt(0);
        transitions.put(fromState + "," + symbol, toState);
    }

    public boolean run(String input) {
        char currentState = startState;
        for (char symbol : input.toCharArray()) {
            if (!alphabet.contains(symbol)) {
                System.out.println("Invalid input symbol: " + symbol);
                return false;
            }
            String key = currentState + "," + symbol;
            if (!transitions.containsKey(key)) {
                System.out.println("No transition for " + key);
                return false;
            }
            currentState = transitions.get(key);
        }
        return acceptStates.contains(currentState);
    }

    public static void main(String[] args) throws IOException {
        Scanner scanner = new Scanner(System.in);
        String filename;

        if (args.length > 0) {
            filename = args[0];
        } else {
            System.out.print("Enter the filename: ");
            filename = scanner.nextLine();
        }

        DFAChecker dfa = new DFAChecker();
        dfa.loadDFA(filename);

        while (true) {
            System.out.print("Enter a string to test (or 'exit' to quit): ");
            String input = scanner.nextLine();
            if (input.equals("exit")) {
                break;
            }
            boolean accepted = dfa.run(input);
            System.out.println("String is " + (accepted ? "accepted" : "rejected") + " by the DFA.");
        }

        scanner.close();
    }
}

File: DFAChecker.py
import re

class DFA:
    def __init__(self):
        self.alphabet = set()
        self.states = set()
        self.start_state = None
        self.accept_states = set()
        self.transitions = {}

    def load_dfa(self, filename):
        with open(filename, 'r') as file:
            lines = file.readlines()
            self.parse_set(lines[0], self.alphabet)
            self.parse_set(lines[1], self.states)
            self.start_state = self.parse_state(lines[2])
            self.parse_set(lines[3], self.accept_states)
            for line in lines[4:]:
                self.parse_transition(line.strip())

    def parse_set(self, line, set_):
        line = re.sub(r'[{}\s]', '', line)
        for char in line:
            set_.add(char)

    def parse_state(self, line):
        return line.strip()[0]

    def parse_transition(self, line):
        parts = line.split('->')
        from_state, symbol = parts[0][1:-1].split(',')
        to_state = parts[1].strip()
        self.transitions[(from_state, symbol)] = to_state

    def run(self, input_str):
        current_state = self.start_state
        for symbol in input_str:
            if symbol not in self.alphabet:
                print(f"Invalid input symbol: {symbol}")
                return False
            if (current_state, symbol) not in self.transitions:
                print(f"No transition for ({current_state}, {symbol})")
                return False
            current_state = self.transitions[(current_state, symbol)]
import re

class DFA:
    def __init__(self):
        self.alphabet = set()
        self.states = set()
        self.start_state = None
        self.accept_states = set()
        self.transitions = {}

    def load_dfa(self, filename):
        with open(filename, 'r') as file:
            lines = file.readlines()
            self.parse_set(lines[0], self.alphabet)
            self.parse_set(lines[1], self.states)
            self.start_state = self.parse_state(lines[2])
            self.parse_set(lines[3], self.accept_states)
            for line in lines[4:]:
                self.parse_transition(line.strip())

    def parse_set(self, line, set_):
        line = re.sub(r'[{}\s]', '', line)
        for char in line:
            set_.add(char)

    def parse_state(self, line):
        return line.strip()[0]

    def parse_transition(self, line):
        parts = line.split('->')
        from_state, symbol = parts[0][1:-1].split(',')
        to_state = parts[1].strip()
        self.transitions[(from_state, symbol)] = to_state

    def run(self, input_str):
        current_state = self.start_state
        for symbol in input_str:
            if symbol not in self.alphabet:
                print(f"Invalid input symbol: {symbol}")
                return False
            if (current_state, symbol) not in self.transitions:
                print(f"No transition for ({current_state}, {symbol})")
                return False
            current_state = self.transitions[(current_state, symbol)]
        return current_state in self.accept_states

def main():
    import sys
    filename = sys.argv[1] if len(sys.argv) > 1 else input("Enter the filename: ")
    dfa = DFA()
    dfa.load_dfa(filename)

    while True:
        input_str = input("Enter a string to test (or 'exit' to quit): ")
        if input_str == 'exit':
            break
        accepted = dfa.run(input_str)
        print(f"String is {'accepted' if accepted else 'rejected'} by the DFA.")

if __name__ == "__main__":
    main()
